### PR TITLE
perf: lower the default allocation for bignums

### PR DESCRIPTION
**Pull request by ggreif**
this addresses the most pressing aspect of #4100, but more improvements are to come

Possible flanking measures:
- if the `HP` still points at the end of the recently allocated bignum after all arithmetic is done, and the bignum's capacity is bigger than its payload, we can cut back the `HP` (this probably only makes sense from specific callsites)
- shrink bignums in the GC — this has been backed out in 3b0eb10de260204c9c6e9f1aedf06e5df4dc9df0, as more work is needed
- Candid conversions for bignums seem cycle-hungry

### CODE REVIEW

File: test/bench/ok/heap-32.drun-run.ok [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ingress Completed: Reply: 0x4449444c016c01b3c4b1f204680100010a00000000000000000101
 ingress Completed: Reply: 0x4449444c0000
-debug.print: (50_227, +91_377_860, 1_433_012_379)
-debug.print: (50_070, +102_586_000, 1_504_942_528)
+debug.print: (50_227, +30_261_252, 666_216_815)
```
Comment (ggreif) [inline test/bench/ok/heap-32.drun-run.ok:3]: @crusso obviously these number came down, should we increase the iteration counts?

Comment (crusso) [reply test/bench/ok/heap-32.drun-run.ok:3]: I dunno, then you'd hide the improvement. Why not leave as is?

Comment (ggreif) [reply test/bench/ok/heap-32.drun-run.ok:3]: I am fine with that, if your original intention wasn't to fill up the heap or similar...

File: rts/motoko-rts/src/types.rs [hunk 2]
```diff
@@ -921,6 +921,10 @@ pub(crate) unsafe fn block_size(address: usize) -> Words<u32> {
 
         TAG_BIGINT => {
             let bigint = address as *mut BigInt;
+            if (*bigint).mp_int.used != (*bigint).mp_int.alloc {
+                println!(100, "used:{}, alloc: {}", (*bigint).mp_int.used, (*bigint).mp_int.alloc);
```
Comment (ggreif) [inline rts/motoko-rts/src/types.rs:925]: Building up the `arr` (wins back 1-2 words)
```
debug.print: used:3, alloc: 5
debug.print: used:5, alloc: 7
debug.print: used:9, alloc: 11
debug.print: used:18, alloc: 19
debug.print: used:36, alloc: 37
debug.print: used:72, alloc: 73
debug.print: used:143, alloc: 145
debug.print: used:286, alloc: 287
debug.print: used:572, alloc: 573
```
Deserialising from Candid blob (wins back 1 word consistently)
```
debug.print: used:3, alloc: 4
debug.print: used:5, alloc: 6
debug.print: used:9, alloc: 10
debug.print: used:18, alloc: 19
debug.print: used:36, alloc: 37
debug.print: used:72, alloc: 73
debug.print: used:143, alloc: 144
debug.print: used:286, alloc: 287
debug.print: used:572, alloc: 573
```

File: rts/motoko-rts/src/types.rs [hunk 3]
```diff
@@ -921,6 +921,10 @@ pub(crate) unsafe fn block_size(address: usize) -> Words<u32> {
...
+            }
+            (*bigint).mp_int.alloc = (*bigint).mp_int.used;
```
Comment (ggreif) [inline rts/motoko-rts/src/types.rs:924]: stats are in https://github.com/dfinity/motoko/pull/4102/files/e15afc367f58f61db4756f6f41e513890d7c8163#r1253257503 and also past commits...

File: rts/motoko-rts/src/bigint.rs [hunk 4]
```diff
@@ -51,6 +51,7 @@ unsafe fn mp_alloc<M: Memory>(mem: &mut M, size: Bytes<u32>) -> *mut u8 {
     let size = size.as_usize();
     debug_assert_eq!((size % core::mem::size_of::<mp_digit>()), 0);
     (*blob).mp_int.alloc = (size / core::mem::size_of::<mp_digit>()) as i32;
+    (*blob).mp_int.used = (*blob).mp_int.alloc; // Otherwise, `block_size() in `allocation_barrier()` may already shorten the object.
```
Comment (ggreif) [inline rts/motoko-rts/src/bigint.rs:54]: The problem with this might be that this changes the value of the bignum (mantissa).

**Review by luc-blaeser — APPROVED** [review]
Thank you very much. This already improves the BigNum size overhead problem substantially.

**Pull request merged**